### ISSUE TITLE
SNI (Server Name Indication) support

### ISSUE THREAD

**Issue by eproxus** [g-andrade/tls_certificate_check#39]
Should this library add the `server_name_indication` option by default? It is never sent by Erlang by default, see the note here: https://www.erlang.org/doc/man/ssl.html#connect-3

General explanation here: https://www.cloudflare.com/en-gb/learning/ssl/what-is-sni/

Hackney, for example sets this by default:

```erlang
1> lists:sort(hackney_connection:ssl_opts("example.com", [])).
[{cacerts,
     [<<48,130,2,9,48,130,1,142,160,3,2,1,2,2,13,2,3,229,192,
        104,239,99,26,156,114,...>>,
      <<48,130,2,9,48,130,1,142,160,3,2,1,2,2,13,2,3,229,184,
        130,235,32,248,37,...>>,
      <<48,130,5,87,48,130,3,63,160,3,2,1,2,2,13,2,3,229,174,
        197,141,4,37,...>>,
      <<48,130,5,87,48,130,3,63,160,3,2,1,2,2,13,2,3,229,147,
        111,49,176,...>>,
      <<48,130,1,220,48,130,1,131,160,3,2,1,2,2,13,2,3,229,126,
        245,63,...>>,
      <<48,130,5,106,48,130,3,82,160,3,2,1,2,2,16,45,221,172,
        206,98,...>>,
      <<48,130,2,27,48,130,1,161,160,3,2,1,2,2,16,65,210,157,
        209,...>>,
      <<48,130,5,86,48,130,3,62,160,3,2,1,2,2,20,67,227,113,...>>,
      <<48,130,2,15,48,130,1,149,160,3,2,1,2,2,20,110,106,...>>,
      <<48,130,6,20,48,130,3,252,160,3,2,1,2,2,8,27,...>>,
      <<48,130,2,84,48,130,1,219,160,3,2,1,2,2,16,...>>,
      <<48,130,5,164,48,130,3,140,160,3,2,1,2,2,...>>,
      <<48,130,5,179,48,130,3,155,160,3,2,1,2,...>>,
      <<48,130,5,192,48,130,3,168,160,3,2,1,...>>,
      <<48,130,2,101,48,130,1,235,160,3,2,...>>,
      <<48,130,5,239,48,130,3,215,160,3,...>>,
      <<48,130,5,130,48,130,3,106,160,...>>,
      <<48,130,2,11,48,130,1,145,...>>,
      <<48,130,5,90,48,130,3,...>>,
      <<48,130,2,110,48,130,...>>,
      <<48,130,5,162,48,...>>,
      <<48,130,2,157,...>>,
      <<48,130,2,...>>,
      <<48,130,...>>,
      <<"0"...>>,<<...>>|...]},
 {customize_hostname_check,
     [{match_fun,#Fun<public_key.6.29392970>}]},
 {depth,100},
 {partial_chain,#Fun<hackney_ssl.0.3524239>},
 {server_name_indication,"example.com"},
 {verify,verify_peer},
 {verify_fun,
     {fun ssl_verify_hostname:verify_fun/3,
      [{check_hostname,"example.com"}]}}]
```

This is the current response from `tls_certificate_check`:
```erlang
2> lists:sort(tls_certificate_check:options("example.com")).
[{cacerts,
     [<<48,130,5,168,48,130,3,144,160,3,2,1,2,2,16,30,211,151,
        9,95,216,180,179,71,112,...>>,
      <<48,130,5,222,48,130,3,198,160,3,2,1,2,2,16,1,253,109,48,
        252,163,202,81,168,...>>,
      <<48,130,3,12,48,130,1,244,160,3,2,1,2,2,3,1,0,32,48,13,6,
        9,42,...>>,
      <<48,130,5,183,48,130,3,159,160,3,2,1,2,2,2,5,9,48,13,6,9,
        42,...>>,
      <<48,130,1,247,48,130,1,125,160,3,2,1,2,2,16,6,37,51,177,
        71,3,...>>,
      <<48,130,5,90,48,130,3,66,160,3,2,1,2,2,16,110,71,169,197,
        75,...>>,
      <<48,130,2,63,48,130,1,197,160,3,2,1,2,2,16,5,85,86,188,
        ...>>,
      <<48,130,3,175,48,130,2,151,160,3,2,1,2,2,16,8,59,224,...>>,
      <<48,130,2,195,48,130,2,74,160,3,2,1,2,2,1,0,48,...>>,
      <<48,130,3,221,48,130,2,197,160,3,2,1,2,2,1,0,...>>,
      <<48,130,3,142,48,130,2,118,160,3,2,1,2,2,16,...>>,
      <<48,130,6,157,48,130,4,133,160,3,2,1,2,2,...>>,
      <<48,130,5,130,48,130,3,106,160,3,2,1,2,...>>,
      <<48,130,3,168,48,130,2,144,160,3,2,1,...>>,
      <<48,130,5,90,48,130,3,66,160,3,2,...>>,
      <<48,130,2,89,48,130,1,223,160,3,...>>,
      <<48,130,4,10,48,130,2,242,160,...>>,
      <<48,130,2,143,48,130,2,21,...>>,
      <<48,130,2,12,48,130,1,...>>,
      <<48,130,5,235,48,130,...>>,
      <<48,130,3,56,48,...>>,
      <<48,130,3,67,...>>,
      <<48,130,5,...>>,
      <<48,130,...>>,
      <<"0"...>>,<<...>>|...]},
 {customize_hostname_check,
     [{match_fun,#Fun<public_key.6.29392970>}]},
 {depth,100},
 {partial_chain,
     fun tls_certificate_check_shared_state:find_trusted_authority/1},
 {verify,verify_peer},
 {verify_fun,
     {fun ssl_verify_hostname:verify_fun/3,
      [{check_hostname,"example.com"}]}}]
```

**Comment by g-andrade** [comment]
Thanks, I missed this one! But I'm confused because the docs for [`ssl:connect/4`](https://www.erlang.org/doc/man/ssl.html#connect-3) say SNI falls back to the hostname, if I'm understanding it correctly:

> When the option server_name_indication is provided, its value (the DNS name) will be used as ReferenceID to [public_key:pkix_verify_hostname/2](https://www.erlang.org/doc/man/public_key.html#pkix_verify_hostname-2). **When no server_name_indication option is given, the Host argument will be used as Server Name Indication extension**.

In any case, best to add test cases to either ensure this supposed default behaviour or to instead realize that it should actually be an explicit option.

**Comment by eproxus** [comment]
Ah odd, the behavior seems to be different between `ssl:connect/3` and `ssl:connect/4`.

@IngelaAndin Would you know the difference between these two APIs?

**Comment by IngelaAndin** [comment]
In case ssl:connect/3 is called with an already existing TCP-socket there is no hostname to fallback on.  Also if an IP-address is uses for the Host argument to connect that will be used which is probably most often is not what you want.

**Comment by g-andrade** [comment]
Changes released in [1.18.0](https://hex.pm/packages/tls_certificate_check/1.18.0)
